jack-core/src/com/rapleaf/jack/DatabaseConnection.java

```diff
@@ -101,21 +101,39 @@ public DatabaseConnection(String dbname_key, long expiration, String driverClass
    * Get a Connection to a database. If there is no connection, create a new one.
    * If the connection hasn't been used in a long time, close it and create a new one.
    * We do this because MySQL has an 8 hour idle connection timeout.
+   *
+   * Because of the potential downtime of CloudSQL, it implements exponential retry policy.
+   * The default retry policy retries five times, which handles SQL downtime less than approx. two minutes.
+   *
    */
 
   public Connection getConnectionInternal() {
-    try {
-      if (conn == null) {
-        Class.forName(driverClass);
-        conn = DriverManager.getConnection(connectionString, username.orNull(), password.orNull());
-      } else if (isExpired() || conn.isClosed()) {
-        resetConnection();
+    int MAX_RETRIES = 7;
+    for (int retryCount = 0; retryCount < MAX_RETRIES; ++retryCount) {
+      try {
+        if  (conn == null) {
+          Class.forName(driverClass);
+          conn = DriverManager.getConnection(connectionString, username.orNull(), password.orNull());
+        } else if (isExpired() || conn.isClosed()) {
+          resetConnection();
+        }
+        updateExpiration();
+        return conn;
+      } catch (Exception e) { //IOEx., SQLEx.
+        // if it is the last retry, throw exception
+        if (retryCount == MAX_RETRIES - 1) {
+          throw new RuntimeException(e);
```

Comment:
I think including some useful error message in this `RuntimeException` will make debugging and/or log parsing a little easier. I'm thinking something along the lines of
```
throw new RuntimeException(
    String.format("Could not establish connection after %d attempts", maxNumRetries),
    e
);
```